### DISCUSSION TITLE
Setting the OutputName property

### DISCUSSION THREAD

**Discussion by MarkStega** [wixtoolset/issues#7317] [Questions] [answered]
In WiX V3 it was pretty common to have something like this fragment to set the name of the output and it's pdb:
```
	<Target Name="BeforeBuild">
		<!-- Get the programs assembly version from the .exe file -->
		<GetAssemblyIdentity AssemblyFiles="..\SystemMonitorService.2012\bin\SystemMonitorService.exe">
			<Output TaskParameter="Assemblies" ItemName="AsmInfo" />
		</GetAssemblyIdentity>
		<!-- Store the assembly version number in ProductVersion preprocessor variable -->
		<CreateProperty Value="$(DefineConstants);ProductVersion=%(AsmInfo.Version)">
			<Output TaskParameter="Value" PropertyName="DefineConstants" />
		</CreateProperty>
		<!-- Name the .msi file after the solution configuration and assembly version e.g TestService-DebugAll-1.4.0.0.msi -->
		<CreateProperty Value="$(SolutionName)-$(Configuration)-%(AsmInfo.Version)">
			<Output TaskParameter="Value" PropertyName="TargetName" />
		</CreateProperty>
		<!-- Name the .wixpdb file after the solution configuration and assembly version e.g TestService-DebugAll-1.4.0.0.msi -->
		<CreateProperty Value="$(TargetName)$(TargetPdbExt)">
			<Output TaskParameter="Value" PropertyName="TargetPdbName" />
		</CreateProperty>
	</Target>
```

I now see in the V4 wixproj that these settings were preserved but a new property called OutputName is defined that drives the output. I tried commenting the propertygroup that defined outputname and then replacing TargetName with OutputName as shown, but I got outputs named after the project name:
```
	<Target Name="BeforeBuild">
		<!-- Get the programs assembly version from the .exe file -->
		<GetAssemblyIdentity AssemblyFiles="..\SystemMonitorService.2012\bin\SystemMonitorService.exe">
			<Output TaskParameter="Assemblies" ItemName="AsmInfo" />
		</GetAssemblyIdentity>
		<!-- Store the assembly version number in ProductVersion preprocessor variable -->
		<CreateProperty Value="$(DefineConstants);ProductVersion=%(AsmInfo.Version)">
			<Output TaskParameter="Value" PropertyName="DefineConstants" />
		</CreateProperty>
		<!-- Name the .msi file after the solution configuration and assembly version e.g TestService-DebugAll-1.4.0.0.msi -->
		<CreateProperty Value="$(SolutionName)-$(Configuration)-%(AsmInfo.Version)">
			<Output TaskParameter="Value" PropertyName="OutputName" />
		</CreateProperty>
	</Target>
```

**Comment by BMurri** [comment, marked as answer]
WiX v3 took a very old version of an MSBuild targets file and adapted it to the needs of the toolset. That resulted in newer features in later versions of Visual Studio, MSBuild, etc. not being available or simply not working when introduced into other projects in the same solution.

WiX v4 integrates into the Microsoft-provided plumbing for SDK integration, taking advantage of all the new features (such as new-style project files, `PackageReference`s, etc.), which means that common properties and targets will now work as they do for other project types using the new-style project file format.

`OutputName` is used to set a lot of other related properties and items early on, before any targets are run, so changing it in a target would also require changing all of the items and properties affected by that property's value at the time your target runs. If your assembly version is generated before your solution is run, you could use an environment variable and set that property as you need it from the get-go.

What I've seen suggested in stackoverflow for other project types in MSBuild is to have a target rename the output file, if the above isn't possible.

**Reply by MarkStega** [reply]
@BMurri 

Thanks, I took the easy way out and set outputname to a known value. After the build I copy the installer & it's pdb using the pattern I desire, then I delete the known name.

So at the top of with wix4 project file:
```
	<PropertyGroup>
		<OutputName>xyzzy</OutputName>
	</PropertyGroup>
```
and at the bottom:
```
	<Import Project="Sdk.targets" Sdk="WixToolset.Sdk" Version="4.0.0-rc.4" />
	
	<Target Name="BeforeBuild">
		<!-- Get the programs assembly version from the .exe file -->
		<GetAssemblyIdentity AssemblyFiles="$(SolutionDir)SystemMonitorService.2012\$(OutputPath)SystemMonitorService.exe">
			<Output TaskParameter="Assemblies" ItemName="AsmInfo" />
		</GetAssemblyIdentity>
		<!-- Store the assembly version number in ProductVersion preprocessor variable -->
		<CreateProperty Value="$(DefineConstants);ProductVersion=%(AsmInfo.Version)">
			<Output TaskParameter="Value" PropertyName="DefineConstants" />
		</CreateProperty>
	</Target>
	
	<Target Name="AfterBuild" AfterTargets="Build">
		<!-- Name the .msi file after the solution configuration and assembly version e.g TestService-DebugAll-1.4.0.0.msi -->
		<CreateProperty Value="$(SolutionName)-$(Configuration)-%(AsmInfo.Version)">
			<Output TaskParameter="Value" PropertyName="NewOutputName" />
		</CreateProperty>
		<Copy SourceFiles="$(ProjectDir)bin\xyzzy.msi" DestinationFiles="$(ProjectDir)bin\$(NewOutputName).msi" />
		<Copy SourceFiles="$(ProjectDir)bin\xyzzy.wixpdb" DestinationFiles="$(ProjectDir)bin\$(NewOutputName).wixpdb" />
		<Delete Files="$(ProjectDir)bin\xyzzy.msi" />
		<Delete Files="$(ProjectDir)bin\xyzzy.wixpdb" />
	</Target>
```
which gives me my desired output of "Optimiser.2012-Debug-2.125.0.0.msi" (and the corresponding wixpdb)
